### ISSUE TITLE
Slow computation on numeric columns 

### ISSUE THREAD

**Issue by 99snowleopards** [vaexio/vaex#1100]
Hello,

 I have an hdf5 file with 400 million rows and 10 columns, all numeric, 9 columns are one of the types uint8, unit32, float32, float 64, the values in these 9 columns have at most 6 digits.

 Only one column is of type unit64, and the values have 10-11 digits

 Doing a groupby and agg (nunique, sum) on any of the 9 columns with < 6 digit values is super fast - takes less than a second.

 The same aggregation (groupby + nunique) on the 10-11 digit value column though takes 90+ seconds - I've also experienced this issue with string columns

 Is there any way to speed up the aggregations on numeric columns with > 10 digit values. I've tried converting to different dtypes, but it takes about the same amount of time.

 Would greatly appreciate any help here

Thanks,

**Comment by maartenbreddels** [comment]
which version of vaex are you using? And can you maybe give some more information, how many unique elements etc

**Comment by 99snowleopards** [comment]
Thanks for the fast response!

vaex: Version: 3.0.0 (should I try upgrading? if so, is this the best way to do it: pip install vaex==4.0.0a2)

the column with 10-11 digits has about 183 mm unique values.

**Comment by maartenbreddels** [comment]
Please do upgrade, `pip install vaex==4.0.0a5` and see if that helps.

How many cores CPU do you have?

**Comment by 99snowleopards** [comment]
looks like upgrading to 4.0.0a5 has helped - on average, the same aggregation is now taking ~60-70 seconds, as opposed to 85 seconds + with 3.0.0

I'm on a powerful EC2 instance - 72 cores (psutil.cpu_count())

is the ~70 second compute time on a numeric column with 10-11 digits (180 mm unique elements) vs 1 second on 6 digit value column (both 400 mm rows) along expected lines? do you see any potential for further improvement here?

**Comment by maartenbreddels** [comment]
For the moment not, but the problem here is that Vaex parallelizes this, while it only gives overhead. For this step you want to reduce the number of threads. Can you try running that step with the env var `VAEX_NUM_THREADS=3`.

I'm working on automating this, so if you can plan with this parameter and let me know you finding, that would be great.

If you want some more background information, let me know!

**Comment by 99snowleopards** [comment]
I played with this variable a little bit - VAEX_NUM_THREADS, set it equal to 3, 24, 36, 72

24 seems to be the optimal number! it's a few seconds faster than 3, so now, its's ~60 seconds for 10-11 digit value column

I'd love some more background information, thanks,

**Comment by maartenbreddels** [comment]
Can I ask why you do a group by on a float column? Did you take a look at
binby or use the binby argument of for instance sum?
Maybe you care to explain what you are trying to do?

(from mobile phone)

**Comment by 99snowleopards** [comment]
I should've specified this in the description: the groupby is on another column (Numeric code - uint8), and the aggregations (nunique, sum...) are done on the other numeric columns, and it is only the aggregation on the 10-11 digit value column that is taking longer.

the output dataframe has about 35 rows.

I haven't tried binby for this - I'll look into that

**Comment by maartenbreddels** [comment]
I thought, that because you talked about decimals, that you grouped by those, and because more decimals means more unique values, that indeed it could be slower.
But are you saying, that two float columns of the same type (float32, or both float64) if summed over in groupby, have different performance?

**Comment by 99snowleopards** [comment]
correct, except that the column that is a lot slower is uint64, it has 11 digit integer values, no floats, an nunique op on this column takes 60 seconds+, as opposed to an nunique on the uint8 column that has integer values, but with fewer than 6 digits in each value (which takes just 1 second), I hope that makes the issue clear - please let me know if I can provide any further info,

**Comment by maartenbreddels** [comment]
Yes that helps, which bits are used on the uint64? Are they all low
numbers, or all high, can you give me a few samples?

(from mobile phone)

**Comment by 99snowleopards** [comment]
any combination of numbers 0-9, example 12345678901
